pkg/client/cache/delta_fifo.go

```diff
@@ -0,0 +1,391 @@
+// AddIfNotPresent().
+//
+// Pop returns a 'Deltas', which has a complete list of all the things
+// that happened to the object (deltas) while it was sitting in the queue.
+func (f *DeltaFIFO) Pop() interface{} {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	for {
+		for len(f.queue) == 0 {
+			f.cond.Wait()
+		}
+		id := f.queue[0]
+		f.queue = f.queue[1:]
+		item, ok := f.items[id]
+		if !ok {
+			// Item may have been deleted subsequently.
+			continue
+		}
+		delete(f.items, id)
+		return item
+	}
+}
+
+// Replace will delete the contents of 'f', using instead the given map.
+// 'f' takes ownersip of the map, you should not reference the map again
+// after calling this function. f's queue is reset, too; upon return, it
+// will contain the items in the map, in no particular order.
+func (f *DeltaFIFO) Replace(list []interface{}) error {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	for _, item := range list {
+		if err := f.queueActionLocked(Sync, item); err != nil {
+			return fmt.Errorf("couldn't enqueue object: %v", err)
+		}
+	}
+	if f.knownObjectKeys == nil {
+		return nil
+	}
+
+	keySet := make(util.StringSet, len(list))
+	for _, item := range list {
+		key, err := f.KeyOf(item)
+		if err != nil {
+			return fmt.Errorf("couldn't create key for object: %#v", err)
+		}
+		keySet.Insert(key)
+	}
+
+	// Detect deletions not already in the queue.
+	knownKeys := f.knownObjectKeys.ListKeys()
+	for _, k := range knownKeys {
+		if _, exists := keySet[k]; exists {
+			continue
+		}
+
+		// This key isn't in the complete set we got, so it must have been deleted.
+		if d, exists := f.items[k]; exists {
+			// Don't issue a delete delta if we have one enqueued as the most
+			// recent delta.
+			if d.Newest().Type == Deleted {
+				continue
+			}
+		}
+		if err := f.queueActionLocked(Deleted, DeletedFinalStateUnknown{k}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// NewDeltaFIFO returns a Store which can be used process changes to items.
+//
+// keyFunc is used to figure out what key an object should have. (It's
+// exposed in the returned DeltaFIFO's KeyOf() method, with bonus features.)
+//
+// 'compressor' may compress as many or as few items as it wants
+// (including returning an empty slice), but it should do what it
+// does quickly since it is called while the queue is locked.
+// 'compressor' may be nil if you don't want any delta compression.
+//
+// 'keyLister' is expected to return a list of keys that the consumer of
+// this queue "knows about". It is used to decide which items are missing
+// when Replace() is called; 'Deleted' deltas are produced for these items.
+// It may be nil if you don't need to detect all deletions.
+// TODO: consider merging keyLister with this object, tracking a list of
+//       "known" keys when Pop() is called. Have to think about how that
+//       affects error retrying.
+func NewDeltaFIFO(keyFunc KeyFunc, compressor DeltaCompressor, knownObjectKeys KeyLister) *DeltaFIFO {
```

Comment:
Move this to the top of the file, since it's arguably the most "external" method?